### PR TITLE
Training recipe of ConvTasNet on LibriMix dataset.

### PR DESCRIPTION
**Pull request by nateanl**

### CODE REVIEW

File: examples/source_separation/lightning_trainer.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    Mapping,
+    List,
+    Optional,
+    Tuple,
+    TypedDict,
+)
+
+import torch
+import torchaudio
+import torchaudio.models
```
Comment (fmassa) [inline examples/source_separation/lightning_trainer.py]: nit: ideally, `torchaudio.models` should be available from `import torchaudio`

File: examples/source_separation/lightning_trainer.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+from pytorch_lightning import LightningModule, Trainer
+from pytorch_lightning.callbacks import ModelCheckpoint, EarlyStopping
+from pytorch_lightning.plugins import DDPPlugin
+from torch import nn
+from torch.optim.lr_scheduler import _LRScheduler
```
Comment (fmassa) [inline examples/source_separation/lightning_trainer.py]: nit for the PyTorch team: this makes me think that we should probably expose a public class for the schedulers, it's generally not good practice to rely on private functions.

cc @datumbox @gchanan for visibility

File: examples/source_separation/lightning_trainer.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+from utils.dataset import utils as dataset_utils
+
+
+class Batch(TypedDict):
+    mix: torch.Tensor  # (batch, time)
+    src: torch.Tensor  # (batch, source, time)
+    mask: torch.Tensor  # (batch, source, time)
+
+
+class SI_SDRi_Metric(nn.Module):
```
Comment (fmassa) [inline examples/source_separation/lightning_trainer.py]: In general, I would have made this a plain function. There is no state hold on here, and now users need to do
```python
SI_SDRi_Metric()(...)
```
to use it.

Another nit: we are mixing `CamelCase` and `snake_case` in the naming, might be good to make it uniform

File: examples/source_separation/lightning_trainer.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    ) -> Tuple[Any]:
+        lr_scheduler = self.lr_scheduler
+        if not lr_scheduler:
+            return self.optim
+        epoch_schedulers = {
+            'scheduler': lr_scheduler,
+            'monitor': 'Losses/val_loss',
+            'interval': 'epoch'
+        }
+        return [self.optim], [epoch_schedulers]
```
Comment (fmassa) [inline examples/source_separation/lightning_trainer.py]: I don't know PyTorchLightning well, but one thing we need to double-check is if `LightningModule` can auto-wrap our given `model` into other abstractions internally (like `DistributedDataParallel`) before calling `configure_optimizers`. If that's the case, then the optimizers / schedulers might point to the unwrapped model, leading to unexpected behavior

File: examples/source_separation/utils/dataset/librimix.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+from pathlib import Path
```
Comment (fmassa) [inline examples/source_separation/utils/dataset/librimix.py:1]: I wonder if this fine should live in `torchaudio.datasets` instead?

File: torchaudio/models/conv_tasnet.py [hunk 6]
```diff
@@ -159,7 +159,7 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
             output = output + skip
         output = self.output_prelu(output)
         output = self.output_conv(output)
-        output = torch.sigmoid(output)
+        output = self.mask_activate(output)
```
Comment (fmassa) [inline torchaudio/models/conv_tasnet.py:170]: This is a change in behavior, is it expected?

**Review by fmassa — COMMENTED** [review]
Thanks for the recipe!

I've done a first round of high-level comments.
For my understanding, did you run the training on multiple GPUs to compare the performance compared to some standard baseline?

File: examples/source_separation/lightning_trainer.py (hunk 3 above)
Comment (nateanl) [reply examples/source_separation/lightning_trainer.py]: Sure. In torchaudio the functional methods follow snake_case, while the transform modules follow CamelCase naming. I will change it to a plain function.

File: examples/source_separation/utils/dataset/librimix.py (hunk 5 above)
Comment (nateanl) [reply examples/source_separation/utils/dataset/librimix.py:1]: Yes, the goal is moving the LibriMix dataset to ``torchaudio.datasets`` at the end. 
I do see such import in current dataset:
https://github.com/pytorch/audio/blob/main/torchaudio/datasets/librispeech.py#L3

File: torchaudio/models/conv_tasnet.py (hunk 6 above)
Comment (nateanl) [reply torchaudio/models/conv_tasnet.py:170]: According to Asteroid implementation, the activation function is relu for LibriMix data, and the performance is better than sigmoid. 
Maybe I can set a parameter to let users choose which activation function to use?

**Comment by nateanl** [comment]
Yes. I'm running the training pipeline to achieve the same performance as Asteroid. Moto tried before but couldn't get the same score, so I'm debugging his implementation to see where I can improve.

**Comment by popcornell** [comment]
Do you guys pad the examples ? We use only slicing in Asteroid and for a good reason. I see here both slicing and padding are implemented. If you pad that is why you can't achieve same performance probably.

**Comment by nateanl** [comment]
> Do you guys pad the examples ? We use only slicing in Asteroid and for a good reason. I see here both slicing and padding are implemented. If you pad that is why you can't achieve same performance probably.

Yes, I reduced the duration of training examples to 3 seconds so that I only use slicing instead of padding. Another change I made is that I use Si-SDR instead of Si-SDR improvement as the loss function. Now I achieved 14.8 dB Si-SDR improvement on sep_clean task :)

**Comment by popcornell** [comment]
Si-SDR and Si-SDR improvement should not make any difference as the gradient is the same.

**Comment by nateanl** [comment]
Yeah, but the Si-SDR improvement may not be stable compared to the Si-SDR loss, that will affect the LR scheduler.

**Comment by popcornell** [comment]
Good point !
